lk2nd/smp/spin-table.c

```diff
@@ -81,17 +81,26 @@ static void smp_spin_table_setup_cpu(struct smp_spin_table *table,
 		dprintf(CRITICAL, "Cannot read reg property of qcom,acc node: %d\n", len);
 		return;
 	}
-	base = fdt32_to_cpu(*val);
+	cpu_power_clamp_base = fdt32_to_cpu(*val);
 
-#if TARGET_MSM8916 || TARGET_MSM8226
-	qcom_power_up_arm_cortex(cpu, base);
-#elif TARGET_MSM8974
 	node = lkfdt_lookup_phandle(fdt, cpu_node, "next-level-cache");
 	if (node < 0) {
 		dprintf(CRITICAL, "Cannot find next-level-cache: %d\n", node);
 		return;
 	}
 
+#if TARGET_MSM8916 || TARGET_MSM8226
+
+	/* Power Domain l2 cache for second cluster */
+	val = fdt_getprop(fdt, node, "reg", &len);
```

Comment:
Is there any existing example for using a memory `reg` in the l2 cache controller node? It does sound reasonable to me, but I am not sure if this will be accepted in mainline. Does this pass `make dtbs_check`?

Reply:
Well, that's the one open question on this.

I don't think "reg" should be a problem since it is a standard property but, I won't know until the DTS gets a first pass upstream review.

The intention is to submit the DTS with 

- enable-method = "spin-table"
- a reference to lk2nd in the commit log and DTS

so that the means of getting all fours cores up is explicit.

The other possibility is to encode the two necessary addresses for the L2 cache inside of lk2nd but, I wasn't sure how you'd feel about that

We could special case it with a compat for msm8936 and msm8939 and it would require no Linux kernel changes to support .. ?

Reply:
If we are open to encoding the address inside of lk2nd somehow, that's one less bit of churn to be submitted and brought through the Linux upstreaming process, might be a good bit easier to do

Reply:
actually let me check what dtbs_check makes of this..

Reply:
I have mixed feelings about this so I can probably be convinced for many different kinds of options. :)

The main reason for taking the addresses from the DT was really just simplicity. I already had the code to parse the CPUs in the DT to set the `cpu-release-addr`. Also, the `qcom,acc` and `qcom,saw` properties were already present in `msm8916.dtsi` upstream because I added support for some devices with (signed) 32-bit-only firmware there.

The difference in your case is that there are no existing bindings for describing the L2 cache memory region. Also, it seems unlikely that anyone will ever want to boot ARM32 Linux on MSM8939 (I think all MSM8939 devices were shipped with a 64-bit TZ firmware). In other words, you would only add the address upstream for lk2nd to use, while there would be no Linux code making use of the additional properties.

I think this is a bit hard to justify, especially considering there is already plenty of Qualcomm code with hardcoded addresses in lk2nd. So, I'm inclined to agree with you that it is not worth the hassle upstream and we should just encode the address in lk2nd for MSM8939.

You could just cleverly derive the base address for the L2 cache from the one given to `qcom_power_up_arm_cortex()` :D

```diff
diff --git a/lk2nd/smp/cpu-boot.c b/lk2nd/smp/cpu-boot.c
index 82b7ffc08..5668dab23 100644
--- a/lk2nd/smp/cpu-boot.c
+++ b/lk2nd/smp/cpu-boot.c
@@ -65,6 +65,9 @@ void qcom_power_up_arm_cortex(uint32_t mpidr, uint32_t base)
 		return;
 	}
 
+	/* Make sure the L2 cache is powered up */
+	qcom_power_up_l2_cache((base & 0xfff00000) + 0x11000);
+
 	pwr_ctl = CPU_PWR_CTL_CLAMP | CPU_PWR_CTL_CORE_MEM_CLAMP |
 		  CPU_PWR_CTL_CORE_RST | CPU_PWR_CTL_COREPOR_RST;
 	writel(pwr_ctl, base + CPU_PWR_CTL);

```

Reply:
`qcom_power_up_l2_cache((base & 0xfff00000) + 0x11000);`

That's nice, I'll steal that code for V2 if you don't mind

Reply:
Sure, maybe add a `#define` for `0x11000` like `APCS_GLB_OFFSET`. Just so it doesn't look like complete magic :)